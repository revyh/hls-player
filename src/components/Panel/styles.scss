@import 'settings.scss';

$panel-height: 27px;

@mixin btn-elem {
  height: $panel-height;
  padding: 1px;
  box-sizing: border-box;
  cursor: pointer;
  fill: currentColor;
}

.base {
  display: flex;
}

.play-btn-elem {
  @include btn-elem;

  width: 23px;
  margin-right: 40px;
}

.mute-btn-elem {
  @include btn-elem;

  width: 27px;
  margin-right: 10px;
}

.mute-btn-elem__active {
  composes: mute-btn-elem;
  padding: 0;
  transform: translateX(2px);
}

.volume-elem {
  display: flex;
  height: $panel-height;
  width: 60px;
  flex-direction: column;
  justify-content: center;
}

.volume-hidden-elem {
  display: none;
}

.volume-track-elem {
  display: block;
  height: 5px;
  background-color: rgba($base-font-color, 0.5);
  border-radius: 5px;
  cursor: pointer;
}

.volume-active-track-elem {
  display: block;
  height: 100%;
  background-color: $base-font-color;
  border-radius: inherit;
  cursor: pointer;
}

.time-elem {
  margin-left: auto;
  margin-right: 40px;
  font-size: 24px;
  line-height: $panel-height;
  visibility: middle;
}

.fullscreen-btn-elem {
  @include btn-elem;

  width: 26px;
}
