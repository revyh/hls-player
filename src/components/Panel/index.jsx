import React from 'react';
import InputRange from 'react-input-range';
import styles from './styles.scss';
import Icon from 'components/Icon';
import getClassName from 'utils/getClassName';
import formatTime from 'utils/formatTime';

const volumeClassNames = {
  inputRange: styles.volumeElem,
  track: styles.volumeTrackElem,
  activeTrack: styles.volumeActiveTrackElem,
  labelContainer: styles.volumeHiddenElem,
  valueLabel: styles.volumeHiddenElem,
  maxLabel: styles.volumeHiddenElem,
  minLabel: styles.volumeHiddenElem,
  slider: styles.volumeHiddenElem,
  sliderContainer: styles.volumeHiddenElem,
};

export default class Panel extends React.Component {
  static defaultProps = {
    isPaused: false,
    isMuted: false,
    isFullscreenEntered: false,
    volume: 50,
    time: 0,
    onVolumeChange() { /* empty */ },
  };

  handlePauseChange = () => {
    const {isPaused, onPauseChange} = this.props;
    onPauseChange && onPauseChange(!isPaused);
  };

  handleMuteChange = () => {
    const {isMuted, onMuteChange} = this.props;
    onMuteChange && onMuteChange(!isMuted);
  };

  handleFullscreenChange = () => {
    const {isFullscreenEntered, onFullscreenChange} = this.props;
    onFullscreenChange && onFullscreenChange(!isFullscreenEntered);
  };

  render() {
    const {
      onVolumeChange,
      isPaused,
      isMuted,
      isFullscreenEntered,
      volume,
      time,
    } = this.props;

    const playSymbol = isPaused
      ? 'play'
      : 'pause';

    const muteSymbol = isMuted
      ? 'mute'
      : 'sound';

    const fullscreenSymbol = isFullscreenEntered
      ? 'leaveFullscreen'
      : 'enterFullscreen';

    const muteClassName = isMuted
      ? styles.muteBtnElemActive
      : styles.muteBtnElem;

    return (
      <div className={getClassName(styles, this.props, styles.base)}>
        <Icon
          className={styles.playBtnElem}
          symbol={playSymbol}
          onClick={this.handlePauseChange}
          key="play"
        />
        <Icon
          className={muteClassName}
          symbol={muteSymbol}
          onClick={this.handleMuteChange}
          key="mute"
        />
        <InputRange
          maxValue={100}
          minValue={0}
          classNames={volumeClassNames}
          value={volume}
          onChange={onVolumeChange}
        />
        <span className={styles.timeElem}>{formatTime(time)}</span>
        <Icon
          className={styles.fullscreenBtnElem}
          symbol={fullscreenSymbol}
          onClick={this.handleFullscreenChange}
          key="fullscreen"
        />
      </div>
    );
  }
}
