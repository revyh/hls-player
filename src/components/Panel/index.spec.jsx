import React from 'react';
import {shallow} from 'enzyme';
import Panel from '.';

jest.mock('components/Icon');

function getButton(element, key) {
  return element.findWhere(
    node => node.name() === 'Icon' && node.key() === key,
  );
}

it('should render with default values', () => {
  expect(shallow(<Panel />)).toMatchSnapshot();
});

it('should render with custom values', () => {
  const element = shallow(
    <Panel
      isPaused
      isMuted
      isFullscreenEntered
      volume={10}
      time={3661}
      onPauseChange={jest.fn()}
      onMuteChange={jest.fn()}
      onFullscreenChange={jest.fn()}
      onVolumeChange={jest.fn()}
    />,
  );

  expect(element).toMatchSnapshot();
});

it('should trigger \'onPauseChange\' event handler', () => {
  const mockOnPauseChange = jest.fn();
  const panel = shallow(
    <Panel isPaused onPauseChange={mockOnPauseChange} />,
  );

  const pauseButton = getButton(panel, 'play');

  pauseButton.simulate('click');

  expect(mockOnPauseChange).toHaveBeenCalledWith(false);
});

it('should trigger \'onMuteChange\' event handler', () => {
  const mockOnMuteChange = jest.fn();
  const panel = shallow(
    <Panel isMuted onMuteChange={mockOnMuteChange} />,
  );

  const muteButton = getButton(panel, 'mute');

  muteButton.simulate('click');

  expect(mockOnMuteChange).toHaveBeenCalledWith(false);
});

it('should trigger \'onFullscreenChange\' event handler', () => {
  const mockOnFullscreenChange = jest.fn();
  const panel = shallow(
    <Panel isFullscreenEntered onFullscreenChange={mockOnFullscreenChange} />,
  );

  const fullscreenButton = getButton(panel, 'fullscreen');

  fullscreenButton.simulate('click');

  expect(mockOnFullscreenChange).toHaveBeenCalledWith(false);
});

it('should trigger \'onVolumeChange\' event handler', () => {
  const mockOnVolumeChange = jest.fn();
  const panel = shallow(
    <Panel volume={10} onVolumeChange={mockOnVolumeChange} />,
  );

  const volumeRange = panel.findWhere(node => node.name() === 'InputRange');

  volumeRange.simulate('change', 40);

  expect(mockOnVolumeChange).toHaveBeenCalledWith(40);
});
