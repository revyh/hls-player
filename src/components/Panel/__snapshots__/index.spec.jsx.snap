// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with custom values 1`] = `
<div
  className="base"
>
  <Icon
    className="playBtnElem"
    onClick={[Function]}
    symbol="play"
  />
  <Icon
    className="muteBtnElemActive"
    onClick={[Function]}
    symbol="mute"
  />
  <InputRange
    classNames={
      Object {
        "activeTrack": "volumeActiveTrackElem",
        "inputRange": "volumeElem",
        "labelContainer": "volumeHiddenElem",
        "maxLabel": "volumeHiddenElem",
        "minLabel": "volumeHiddenElem",
        "slider": "volumeHiddenElem",
        "sliderContainer": "volumeHiddenElem",
        "track": "volumeTrackElem",
        "valueLabel": "volumeHiddenElem",
      }
    }
    disabled={false}
    maxValue={100}
    minValue={0}
    onChange={[Function]}
    step={1}
    value={10}
  />
  <span
    className="timeElem"
  >
    01:01:01
  </span>
  <Icon
    className="fullscreenBtnElem"
    onClick={[Function]}
    symbol="leaveFullscreen"
  />
</div>
`;

exports[`should render with default values 1`] = `
<div
  className="base"
>
  <Icon
    className="playBtnElem"
    onClick={[Function]}
    symbol="pause"
  />
  <Icon
    className="muteBtnElem"
    onClick={[Function]}
    symbol="sound"
  />
  <InputRange
    classNames={
      Object {
        "activeTrack": "volumeActiveTrackElem",
        "inputRange": "volumeElem",
        "labelContainer": "volumeHiddenElem",
        "maxLabel": "volumeHiddenElem",
        "minLabel": "volumeHiddenElem",
        "slider": "volumeHiddenElem",
        "sliderContainer": "volumeHiddenElem",
        "track": "volumeTrackElem",
        "valueLabel": "volumeHiddenElem",
      }
    }
    disabled={false}
    maxValue={100}
    minValue={0}
    onChange={[Function]}
    step={1}
    value={50}
  />
  <span
    className="timeElem"
  >
    00:00
  </span>
  <Icon
    className="fullscreenBtnElem"
    onClick={[Function]}
    symbol="enterFullscreen"
  />
</div>
`;
